### PR TITLE
Add ability to filter tests on the command-line

### PR DESCRIPTION
**Pull request by janko-m**
Now we can run something like

```
$ ruby -r test/something_test.rb 'something'
```

and it will run all tests matching that regex.

### CODE REVIEW

File: lib/protest/test_case.rb [hunk 1]
```diff
@@ -32,7 +32,9 @@ class TestCase
     # Run all tests in this context. Takes a Runner instance in order to
     # provide output.
     def self.run(runner)
-      tests.each {|test| runner.report(test) }
+      regex = Regexp.new(ARGV.first.to_s)
+      tests_to_run = tests.select { |test| "#{test.class.description} #{test.name}" =~ regex }
```
Comment (inkel) [inline lib/protest/test_case.rb]: I'd create the regexp and filter the tests only if `ARGV.first` is passed, and maybe make it a flag? Like `-m "regexp"`.

**Comment by janko-m** [comment]
Probably would be good to include line number as well, we're already saving the information where is the test defined. I will add that tomorrow.

I wanted to make this minimalistic, is it better to maybe make it a CLI option using Ruby's `optparse` stdlib?

**Comment by matflores** [comment]
> Probably would be good to include line number as well, we're already saving the information where is the test defined. I will add that tomorrow.

@janko-m good idea, but please let's open a separate PR for that :)

> I wanted to make this minimalistic, is it better to maybe make it a CLI option using Ruby's optparse stdlib?

Ideally it should be supported regardless of how you run your tests. Adding a CLI tool for Protest should be done on a separate PR as well, I wouldn't add it here.

**Comment by janko-m** [comment]
@matflores I was actually thinking to just detect command-line arguments when running tests regularly, without introducing a CLI tool. The `ruby` executable supports that, and it's what Minitest is doing as well. I updated the PR to use the following notation:

```rb
ruby -r test/something_test.rb --name 'something'
ruby -r test/something_test.rb -n 'something'
```

Afterwards I will add the line numbers as a `-l` or `--line` option, so that it's super-easy to run a single test.

**Comment by jemc** [comment]
If you wanted to avoid CLI arg parsing, you could use an environment variable - protest already has some `PROTEST_*` env vars that it checks, if I recall correctly.

**Comment by asterite** [comment]
Bump :-)

Is there anything missing to merge this?

I'm also curious: what's the benefit of using Protest instead of RSpec?

**Comment by janko-m** [comment]
I think the only thing needed is to switch from CLI argument to environment variable. But I would also actually like to switch it to support line numbers instead, I can work on it next week when I'm back from vacation.

Instead od comparing Protest with RSpec, I think it makes more sense to compare it with Minitest. I for one prefer Minitest over RSpec (some time ago it was the opposite), and I didn't see advantages of Protest over Minitest -- the former seems to me like a subset of the latter in terms of features. Minitest has more assertions, nearest test feature, `minitest-line` plugin for using line numbers instead of regexes, other plugins etc.

However, Joe had an issue with Minitest related to threading (I think it was something really simple with threads that didn't work), and author didn't think it should be fixed, even though Joe had offered a solution in form of a PR. And in my experience the Minitest maintainer isn't nice (unlike the RSpec maintainer who is very nice).

So, my overly simplified view is that we have Protest which is simplest and can be maintained by us, Minitest which has more features and is still simple but doesn't always have best decisions, and RSpec which has the most features out-of-the box though it's also more complex.

**Comment by matflores** [comment]
> I'm also curious: what's the benefit of using Protest instead of RSpec?

@asterite having a much simpler tool that covers most of the use cases I needed. At the beginning speed was another one, although I admit I haven’t benchmarked it against RSpec lately.

> I think the only thing needed is to switch from CLI argument to environment variable. But I would also actually like to switch it to support line numbers instead, I can work on it next week when I'm back from vacation.

Even if the ENV approach would be easier to implement, I prefer to use environment variables for permanent settings and rely on args for one time settings like in this case. But parsing of args shouldn't happen inside `run`, on the implementation I had in mind `run` was going to receive all the options it needs to run the test suite, and the parsing would occur on a different place before calling `run` (ideally in a CLI tool, but that I've never implemented it).

@janko-m please let's keep the discussions focused on tools and not people here. Thanks!

**Comment by janko-m** [comment]
> @janko-m please let's keep the discussions focused on tools and not people here. Thanks!

:+1:

One more thing that I didn't find ideal in Minitest is that it uses threads to run examples, even when not using `parallelize_me!`. That way I cannot easily use RubyProf to find out what's making a test suite slow, because each test run is in a new thread, so RubyProf output isn't useful. Am I correct that Protest doesn't use threads?

> But parsing of args shouldn't happen inside run, on the implementation I had in mind run was going to receive all the options it needs to run the test suite, and the parsing would occur on a different place before calling run (ideally in a CLI tool, but that I've never implemented it).

I can do it this way. Passing script-level CLI arguments to `ruby` has always been awkward for me.

**Comment by asterite** [comment]
Bump again :-)

I think the current implementation is fine. I wouldn't use ENV vars because it would be impossible to select a test case when running inside docker: I'd have to change the docker-compose file... so using a command line argument seems fine.

Specifying a line number would be nice too, but could be done in a separate PR.

**Comment by jemc** [comment]
> I wouldn't use ENV vars because it would be impossible to select a test case when running inside docker: I'd have to change the docker-compose file...

It's not really impossible - you can use the `env` command to accomplish this - `docker exec -ti mycontainer env VAR1=value scripts/test-run`

**Comment by asterite** [comment]
@jemc Oh, nice! I still think using env vars for this is a bit overkill :-)

**Comment by janko-m** [comment]
I will close this issue, as I decided to implement line numbers directly instead of regex filters, because that's what I actually needed. If anyone wants to use regex filters, they can open another PR to add that in.

**Pull request closed without merging**